File: system/arcstandard.go
// Copyright 2015 The dpar Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package system

import (
	"errors"
	"fmt"
	"strings"
)

// Archetype transitions + interface validation.
var archetypeASShift Transition = asShift{}
var archetypeASLeftArc Transition = asLeftArc{"<archetype>"}
var archetypeASRightArc Transition = asRightArc{"<archetype>"}

// Assert TransitionSystem/TransitionSerializer conformance.
var _ TransitionSystem = NewArcStandard()
var _ TransitionSerializer = NewArcStandard()

// The ArcStandard transition system.
type ArcStandard struct {
	archetypeTransitions TransitionSet
}

// NewArcStandard creates a new arc-standard transition system.
func NewArcStandard() *ArcStandard {
	trans := map[Transition]interface{}{
		archetypeASShift:    nil,
		archetypeASLeftArc:  nil,
		archetypeASRightArc: nil,
	}

	return &ArcStandard{trans}
}

// IsTerminal checks whether a parser configuration is a terminal
// configuration for the system. When this is the case, no further
// transitions should be attempted.
func (ts *ArcStandard) IsTerminal(c Configuration) bool {
	return len(c.Buffer) == 0
}

// PossibleTransitions returns the set of possible transitions in a
// particular parser configuration. The returned transitions are
// archetypal transitions (they do not parametrized over a dependency
// relation).
func (ts *ArcStandard) PossibleTransitions(configuration Configuration) TransitionSet {
	possible := make(TransitionSet)

	for trans := range ts.archetypeTransitions {
		if trans.IsPossible(configuration) {
			possible[trans] = nil
		}
	}

	return possible
}

// SerializeTransition serializes a transition of the arc-standard
// system to a string. An error is returned when it is attempted to
// serialize a transition of another system.
func (ts *ArcStandard) SerializeTransition(t Transition) (string, error) {
	switch t := t.(type) {
	case asLeftArc:
		return fmt.Sprintf("LEFT_ARC %s", t.relation), nil
	case asRightArc:
		return fmt.Sprintf("RIGHT_ARC %s", t.relation), nil
	case asShift:
		return "SHIFT", nil
	default:
		return "", fmt.Errorf("Not a transition of the arc standard system: %T", t)
	}
}

// DeserializeTransition deserializes a transition from the arc-standard system.
// If the provided transition is not known to the system, an error is returned.
func (ts *ArcStandard) DeserializeTransition(transString string) (Transition, error) {
	parts := strings.SplitN(transString, " ", 2)

	if len(parts) == 0 {
		return nil, errors.New("Empty transition")
	}

	switch parts[0] {
	default:
		return nil, fmt.Errorf("Unknown transition: %s", parts[0])
	case "LEFT_ARC":
		if len(parts) == 1 {
			return nil, errors.New("Left-Arc transition requires label argument")
		}
		return asLeftArc{parts[1]}, nil
	case "RIGHT_ARC":
		if len(parts) == 1 {
			return nil, errors.New("Right-Arc transition requires label argument")
		}
		return asRightArc{parts[1]}, nil
	case "SHIFT":
		return asShift{}, nil
	}
}

type asLeftArc struct {
	relation string
}

func (l asLeftArc) IsPossible(c Configuration) bool {
	stackSize := len(c.Stack)
	return len(c.Stack) != 0 && len(c.Buffer) != 0 &&
		c.Stack[stackSize-1] != 0
}

func (l asLeftArc) Apply(c *Configuration) {
	stackSize := len(c.Stack)
	head := c.Buffer[0]
	dependant := c.Stack[stackSize-1]
	dependency := Dependency{head, l.relation, dependant}

	c.AddDependency(&dependency)
	c.Stack = c.Stack[:stackSize-1]
}

type asRightArc struct {
	relation string
}

func (r asRightArc) IsPossible(c Configuration) bool {
	return len(c.Stack) != 0 && len(c.Buffer) != 0
}

func (r asRightArc) Apply(c *Configuration) {
	stackSize := len(c.Stack)
	dependant := c.Buffer[0]
	head := c.Stack[stackSize-1]
	dependency := Dependency{head, r.relation, dependant}

	c.AddDependency(&dependency)
	c.Stack = c.Stack[:stackSize-1]
	c.Buffer[0] = head
}

type asShift struct{}

func (s asShift) IsPossible(c Configuration) bool {
	return len(c.Buffer) != 0
}

func (s asShift) Apply(c *Configuration) {
	token := c.Buffer[0]
	c.Buffer = c.Buffer[1:]
	c.Stack = append(c.Stack, token)
}

// The ArcStandardOracle returns a correct transition for a particular
// configuration in the arc-standard system. The transition choice in
// an oracle is based on the gold-standard parse.
type ArcStandardOracle struct {
	dependentHeadMapping map[uint]Dependency
}

// NewArcStandardOracle returns a new oracle from the arc-standard system.
// The gold dependency set is used to choose a correct transition from
// each configuration.
func NewArcStandardOracle(goldDependencies DependencySet) Guide {
	oracle := ArcStandardOracle{goldDependencies.CreateDependentHeadMapping()}
	return &oracle
}

// BestTransition returns the best transition from the current
// configuration, as inferred from the gold standard dependency
// structure.
func (o *ArcStandardOracle) BestTransition(c *Configuration) Transition {
	stackSize := len(c.Stack)
	if stackSize != 0 {
		stackTip := c.Stack[stackSize-1]
		bufferHead := c.Buffer[0]

		la := asLeftArc{o.dependentHeadMapping[stackTip].Relation}
		if la.IsPossible(*c) && o.dependentHeadMapping[stackTip].Head == bufferHead {
			return la
		}

		ra := asRightArc{o.dependentHeadMapping[bufferHead].Relation}
		if ra.IsPossible(*c) && o.dependentHeadMapping[bufferHead].Head == stackTip &&
			!o.neededForAttachment(c, bufferHead) {
			return ra
		}
	}

	return asShift{}
}

func (o *ArcStandardOracle) neededForAttachment(c *Configuration, token uint) bool {
	for _, bufToken := range c.Buffer {
		if o.dependentHeadMapping[bufToken].Head == token {
			return true
		}
	}

	return false
}

File: dpar/src/systems/arc_standard.rs
use std::collections::HashMap;

use guide::Guide;
use system::{
    Dependency, DependencySet, ParserState, Transition, TransitionLookup, TransitionSystem,
};

use features::addr::Source;
use system::AttachmentAddr;
use systems::util::dep_head_mapping;

#[derive(Eq, PartialEq, Serialize, Deserialize)]
pub struct ArcStandardSystem {
    transitions: TransitionLookup<ArcStandardTransition>,
}

impl ArcStandardSystem {
    pub fn new() -> Self {
        ArcStandardSystem {
            transitions: TransitionLookup::default(),
        }
    }
}

impl Default for ArcStandardSystem {
    fn default() -> Self {
        ArcStandardSystem::new()
    }
}

impl TransitionSystem for ArcStandardSystem {
    type Transition = ArcStandardTransition;
    type Oracle = ArcStandardOracle;

    const ATTACHMENT_ADDRS: [AttachmentAddr; 2] = [
        AttachmentAddr {
            head: Source::Buffer(0),
            dependent: Source::Stack(0),
        },
        AttachmentAddr {
            head: Source::Stack(0),
            dependent: Source::Buffer(0),
        },
    ];

    fn is_terminal(state: &ParserState) -> bool {
        state.buffer().is_empty()
    }

    fn oracle(gold_dependencies: &DependencySet) -> Self::Oracle {
        ArcStandardOracle::new(gold_dependencies)
    }

    fn transitions(&self) -> &TransitionLookup<Self::Transition> {
        &self.transitions
    }
}

/// Arc-standard transition.
#[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
pub enum ArcStandardTransition {
    LeftArc(String),
    RightArc(String),
    Shift,
}

impl Transition for ArcStandardTransition {
    type S = ArcStandardSystem;

    fn is_possible(&self, state: &ParserState) -> bool {
        match self {
            &ArcStandardTransition::LeftArc(_) => {
                !state.stack().len() > 1 && !state.buffer().is_empty()
            }
            &ArcStandardTransition::RightArc(_) => {
                !state.stack().is_empty() && !state.buffer().is_empty()
            }
            &ArcStandardTransition::Shift => !state.buffer().is_empty(),
        }
    }

    fn apply(&self, state: &mut ParserState) {
        let stack_size = state.stack().len();

        match self {
            &ArcStandardTransition::LeftArc(ref rel) => {
                let head = state.buffer()[0];
                let dependent = state.stack()[stack_size - 1];

                state.add_dependency(Dependency {
                    head: head,
                    relation: rel.clone(),
                    dependent: dependent,
                });

                state.stack_mut().pop();
            }
            &ArcStandardTransition::RightArc(ref rel) => {
                let dependent = state.buffer()[0];
                let head = state.stack()[stack_size - 1];

                state.add_dependency(Dependency {
                    head: head,
                    relation: rel.clone(),
                    dependent: dependent,
                });

                state.buffer_mut()[0] = state.stack()[stack_size - 1];
                state.stack_mut().pop();
            }
            &ArcStandardTransition::Shift => {
                let next = state.buffer()[0];
                state.stack_mut().push(next);
                state.buffer_mut().remove(0);
            }
        }
    }
}

/// Arc-standard system oracle.
pub struct ArcStandardOracle {
    dependencies: HashMap<usize, Dependency>,
}

impl ArcStandardOracle {
    pub fn new(gold_dependencies: &DependencySet) -> ArcStandardOracle {
        ArcStandardOracle {
            dependencies: dep_head_mapping(gold_dependencies),
        }
    }

    fn needed_for_attachment(&self, state: &ParserState, token: usize) -> bool {
        for buf_token in state.buffer() {
            if let Some(dep) = self.dependencies.get(&buf_token) {
                if dep.head == token {
                    return true;
                }
            }
        }

        false
    }
}

impl Guide for ArcStandardOracle {
    type Transition = ArcStandardTransition;

    fn best_transition(&mut self, state: &ParserState) -> ArcStandardTransition {
        let stack = &state.stack();
        let buffer = &state.buffer();

        if !stack.is_empty() {
            let stack_tip = stack[stack.len() - 1];
            let buffer_head = buffer[0];

            if let Some(dep) = self.dependencies.get(&stack_tip) {
                let la = ArcStandardTransition::LeftArc(dep.relation.clone());

                if dep.head == buffer_head && la.is_possible(state) {
                    return la;
                }
            }

            if let Some(dep) = self.dependencies.get(&buffer_head) {
                let ra = ArcStandardTransition::RightArc(dep.relation.clone());

                if dep.head == stack_tip
                    && ra.is_possible(state)
                    && !self.needed_for_attachment(state, buffer_head)
                {
                    return ra;
                }
            }
        }

        ArcStandardTransition::Shift
    }
}
